packages/core/src/browser/shell/application-shell.ts

```diff
@@ -636,6 +636,12 @@ export class ApplicationShell extends Widget {
         // Proceed with the main panel once all others are set up
         await this.bottomPanelState.pendingUpdate;
         if (mainPanel) {
+            // Array of widgets must not contain items with `undefined` value as it breaks the restoring of the layout
```

Comment:
T.b.h I liked the previous version better: can't we just filter the array in `ShellLayoutRestorer#parse()` after we've created all the widgets?

Reply:
And I like the old version better because the filtering is done closer to where the widget is created (or not created). Simpler to understand when looking at ShellLayoutRestorere.

Reply:
This what I wanted to do first, but in this case it will not work.
In `ShellLayoutRestorer#parse()` each widget is created in a separate promise, which is called already after returning array of widgets.

https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L272-L279

Reply:
We wait can for all promises to be fullfilled: we should not return before the `widgets` array is fully populated anyway:

Reply:
In previous version we had a problem with order of opened editors. It's because the time for restoring/creating widgets for different files is different and the created widget was always inserted the first.

Layout data is created in `ShellLayoutRestorer#inflate` function. 
At the beginning of this function we parse layout string representation to have `LayoutData`.
https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L228

But at that moment arrays of widgets are empty for all panels.

At the end we call `parseContext.inflate(context)`. It runs all the promises, which create the widgets and insert them in the array.

https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L256

Reply:
Everything goes in  `ShellLayoutRestorer#inflate`

The first we parse the layout, but without creating the widget instances.
Then we check for the layout version and perform some migrations.
And only then if everything's fine we run `parseContext.inflate(context)` and create the widgets.
Here the idea was do not create widgets before checks.

Reply:
That's why to preserve order of opened editors we insert the widget to the array by index.

https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L276

Reply:
Yes, but we can do 
```
for (let i = 0; i < descs.length; i++) {
  const promises= [];
  parseContext.push(async context => {
     promises.push(this.convertToWidget(descs[i], context).then(widget => widgets[i] = widget));
});}
await Promise.all(promises);
return widgets.filter(widget => widget !== undefined);
```

Reply:
Yes, it would be great to do this filtering directly where the arrays are filled, not where they are consumed.

Reply:
@tsmaeder @spoenemann
If we create a widgets in the `ShellLayoutRestorer#parse` method by adding `await Promise.all(promises)`, then the code in `ShellLayoutRestorer#inflate` does not have a big sense.

The idea was:
- prepare `ShellLayoutRestorer.ParseContext` (array of promises)
- parse the layout without creating the widgets, but add promises to `ParseContext` for each widget
- perform some checks and migrations
- create widgets with `await parseContext.inflate(context)`
https://github.com/eclipse-theia/theia/blob/b54e9eb75ca6883fa2a4bc0c1569878132798cd2/packages/core/src/browser/shell/shell-layout-restorer.ts#L256

Yes, it will be much clear if we do everything when parsing the JSON in `ShellLayoutRestorer#parse`. But again, then we need to completely rework `ShellLayoutRestorer#inflate`. 

@akosyakov would you like to comment? I see you introduced this code a year ago
https://github.com/eclipse-theia/theia/commit/f8a333ff025483c10fe57a79ed83a5d5173b888c#diff-5974c221773ca89483f3da1843aa00038cf2d6b0eb398fce80d3f31434f2e1ebR212

Reply:
Would it work to move the filtering code to the end of `ShellLayoutRestorer#inflate`, just before returning the layout data?

Reply:
Why do we need to change ShellLayouRestorer#parse if no null values are ever returned? Just remove that change!

Reply:
I thought we had a change originally that worked, it just did not respect the order. Return to that change and fix the ordering.

Reply:
> Would it work to move the filtering code to the end of `ShellLayoutRestorer#inflate`, just before returning the layout data?

I think it will work, but the check will be only for specific panel
```
    if (layout.mainPanel && layout.mainPanel.main && layout.mainPanel.main.type === 'tab-area') {
        const config = layout.mainPanel.main as DockLayout.ITabAreaConfig;
        config.widgets = config.widgets.filter(value => value !== undefined);
    }
```

Reply:
> I thought we had a change originally that worked, it just did not respect the order. Return to that change and fix the ordering.

Do you mean a change in https://github.com/eclipse-theia/theia/pull/8621 ?
At the first glance it looks nice. But the problem with ordering was because the time for restoring different editors with different type of files was different. There is no guarantee that promises that create widgets will be resolved in predefined sequence.
I wrote about that in my comment
https://github.com/eclipse-theia/theia/pull/8621#issuecomment-717889244

Reply:
Yes, that change. The code I proposed in https://github.com/eclipse-theia/theia/pull/8680/files#r520333442 takes care of the ordering problem.

Reply:
The "not storing the state if there is not serializer" still makes sense, though.

Reply:
> Yes, that change. The code I proposed in https://github.com/eclipse-theia/theia/pull/8680/files#r520333442 takes care of the ordering problem.

Yes, it takes. But then widgets will be created before checks and migrations, that are in `ShellLayoutRestorer#inflate`

Reply:
@tsmaeder @spoenemann I have another idea is to move filtering at the end of `ParseContext#inflate`.

```
    export class ParseContext {
        protected readonly toInflate: Inflate[] = [];
        protected readonly arrays: WidgetArray[] = [];

        newArray(): WidgetArray {
            const array: WidgetArray = [];
            this.arrays.push(array);
            return array;
        }

        async inflate(context: InflateContext): Promise<void> {
            const pending: Promise<void>[] = [];
            while (this.toInflate.length) {
                pending.push(this.toInflate.pop()!(context));
            }
            await Promise.all(pending);

            this.arrays.forEach(array => array = array.filter(value => value !== undefined));
        }
    }
```

Then `ShellLayoutRestorer#parse` can ask for a widget array, which will be filtered after creating the widgets
```
    const widgets = parseContext.newArray();
    const descs = (value as WidgetDescription[]);
    for (let i = 0; i < descs.length; i++) {
        parseContext.push(async context => {
            widgets[i] = await this.convertToWidget(descs[i], context);
        });
    }
    return widgets;
```

Reply:
@tsmaeder @spoenemann I have another idea is to move filtering at the end of `ShellLayoutRestorer.ParseContext#inflate`.

But the `ParseContext` must have all the arrays to be filtered.

```
    export class ParseContext {
        protected readonly toInflate: Inflate[] = [];
        protected readonly toFilter: Widgets[] = [];

        newArray(): Widgets {
            const array: Widgets = [];
            this.toFilter.push(array);
            return array;
        }

        async inflate(context: InflateContext): Promise<void> {
            const pending: Promise<void>[] = [];
            while (this.toInflate.length) {
                pending.push(this.toInflate.pop()!(context));
            }
            await Promise.all(pending);

            if (this.toFilter.length) {
                this.toFilter.forEach(array => {
                    const filtered = array.filter(value => value !== undefined);
                    array.splice(0, array.length, ...filtered);
                });
            }
        }
    }

    export type WidgetArray = (Widget | undefined)[];
```

Then `ShellLayoutRestorer#parse` can ask for a widget array, which will be filtered after creating the widgets
```
    const widgets = parseContext.newArray();
    const descs = (value as WidgetDescription[]);
    for (let i = 0; i < descs.length; i++) {
        parseContext.push(async context => {
            widgets[i] = await this.convertToWidget(descs[i], context);
        });
    }
    return widgets;
```

I have just checked this approach, it works well.

Reply:
Sounds good.

Reply:
Reworked

Reply:
I really don't understand why this needs to be so complicated: just wait for all the promises to resolve and then filter out the undefined values from the array in ShellLayoutRestorer#parse(). It's better code-quality-wise because it does the filtering closer to where the nulls are created and there is no if conditional expression around the filtering. Please convince me!

Reply:
It's such confusing because promises for the widgets are created in one place and then resolved in another. 
Why do we need so?
It could happen that your Theia is updated or downgraded and the layout cannot be restored properly.
That requires to do some checks for the layout version and probably transform the layout or update the state of some widgets before creating their instances.
Just take a look at `ShellLayoutRestorer#inflate()`